src/main/java/vazkii/patchouli/client/book/page/abstr/PageDoubleRecipeRegistry.java

```diff
@@ -10,24 +10,24 @@
 import vazkii.patchouli.client.book.BookContentsBuilder;
 import vazkii.patchouli.client.book.BookEntry;
 import vazkii.patchouli.common.base.Patchouli;
-import vazkii.patchouli.mixin.AccessorRecipeManager;
 
 import javax.annotation.Nullable;
 
-import java.util.Map;
-
 public abstract class PageDoubleRecipeRegistry<T extends Recipe<?>> extends PageDoubleRecipe<T> {
 	private final RecipeType<? extends T> recipeType;
 
 	public PageDoubleRecipeRegistry(RecipeType<? extends T> recipeType) {
 		this.recipeType = recipeType;
 	}
 
+	@SuppressWarnings("unchecked")
 	@Nullable
 	private T getRecipe(ResourceLocation id) {
+		if (Minecraft.getInstance().level == null) {
+			return null;
+		}
 		RecipeManager manager = Minecraft.getInstance().level.getRecipeManager();
-		Map<ResourceLocation, T> recipes = (Map<ResourceLocation, T>) ((AccessorRecipeManager) manager).patchouli_byType(recipeType);
-		return recipes.get(id);
+		return (T) manager.byKey(id).filter(recipe -> recipe.getType() == recipeType).orElse(null);
```

Comment:
this is not equivalent.

Reply:
1. actually it is equivalent (it gets the recipe and validates its type vs it gets the map for the type and gets the recipe)
2. no
3. will do

Reply:
are you sure? in `byKey` I see that vanilla maps then performs a `findFirst`, which makes me incredibly suspicious.

Reply:
findFirst is just the main way to convert a stream to an optional

Reply:
there can only be one recipe by id because all recipes that are in the recipe manager are under the same folder